File: jobx-backend/src/main/java/com/jobx/util/TextLists.java
package com.jobx.util;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

/**
 * Normalization for user-entered word lists (FilterProfile keywords and
 * exclude words): trim entries, drop blanks, dedupe case-insensitively while
 * preserving first-seen order and original casing.
 */
public final class TextLists {

    private TextLists() {
    }

    public static List<String> normalize(List<String> raw) {
        if (raw == null) {
            return List.of();
        }
        List<String> result = new ArrayList<>();
        Set<String> seen = new HashSet<>();
        for (String entry : raw) {
            if (entry == null) {
                continue;
            }
            String trimmed = entry.trim();
            if (trimmed.isEmpty()) {
                continue;
            }
            if (seen.add(trimmed.toLowerCase(Locale.ROOT))) {
                result.add(trimmed);
            }
        }
        return result;
    }
}

File: jobx-frontend/src/app/core/util/text-lists.ts
/**
 * Mirrors the backend's com.jobx.util.TextLists.normalize: trim, drop blanks,
 * dedupe case-insensitively, keep first-seen order and original casing.
 *
 * Doing it client-side too means the preferences form shows the user exactly
 * what the server will store, and the "at least one keyword" check fails in the
 * form instead of as a 400 round trip.
 */
export function normalizeList(values: readonly string[]): string[] {
  const out: string[] = [];
  const seen = new Set<string>();
  for (const raw of values) {
    const value = (raw ?? '').trim();
    if (!value) continue;
    const key = value.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    out.push(value);
  }
  return out;
}

/** Comma-separated textarea -> normalized list. */
export function parseList(input: string): string[] {
  return normalizeList((input ?? '').split(','));
}
